src/category_theory/shift.lean

```diff
@@ -1,46 +1,346 @@
+
+/-- Shifting by `i + j` is the same as shifting by `i` and then shifting by `j`. -/
+abbreviation shift_functor_add (i j : A) :
+  shift_functor C (i + j) ≅ shift_functor C i ⋙ shift_functor C j :=
+((shift_monoidal_functor C A).μ_iso i j).symm
+
+variables (A)
+
+/-- Shifting by zero is the identity functor. -/
+abbreviation shift_functor_zero : shift_functor C (0 : A) ≅ 𝟭 C :=
+(shift_monoidal_functor C A).ε_iso.symm
 
 -- Any better notational suggestions?
-notation X`⟦`n`⟧`:20 := ((shift _)^(n : ℤ)).functor.obj X
-notation f`⟦`n`⟧'`:80 := ((shift _)^(n : ℤ)).functor.map f
+notation X`⟦`n`⟧`:20 := (shift_functor _ n).obj X
+notation f`⟦`n`⟧'`:80 := (shift_functor _ n).map f
+
+end defs
+
+section examples
+variables [has_shift C ℤ]
+
+example {X Y : C} (f : X ⟶ Y) : X⟦(1 : ℤ)⟧ ⟶ Y⟦1⟧ := f⟦1⟧'
+example {X Y : C} (f : X ⟶ Y) : X⟦(-2 : ℤ)⟧ ⟶ Y⟦-2⟧ := f⟦-2⟧'
+
+end examples
+
+section add_monoid
+
+variables {C A} [add_monoid A] [has_shift C A] (X Y : C) (f : X ⟶ Y)
+
+@[simp] lemma has_shift.shift_app (n : A) (X : C) : (has_shift.shift.obj n).obj X = X⟦n⟧ := rfl
+
+/-- Shifting by `i + j` is the same as shifting by `i` and then shifting by `j`. -/
+abbreviation shift_add (i j : A) : X⟦i + j⟧ ≅ X⟦i⟧⟦j⟧ := (shift_functor_add C i j).app _
+
+@[reassoc] lemma shift_add_hom_comp_eq_to_hom₁ (i i' j : A) (h : i = i') :
+  (shift_add X i j).hom ≫ eq_to_hom (by rw h) = eq_to_hom (by rw h) ≫ (shift_add X i' j).hom :=
+by { cases h, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma shift_add_hom_comp_eq_to_hom₂ (i j j' : A) (h : j = j') :
+  (shift_add X i j).hom ≫ eq_to_hom (by rw h) = eq_to_hom (by rw h) ≫ (shift_add X i j').hom :=
+by { cases h, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma shift_add_hom_comp_eq_to_hom₁₂ (i j i' j' : A) (h₁ : i = i') (h₂ : j = j') :
+  (shift_add X i j).hom ≫ eq_to_hom (by rw [h₁, h₂]) =
+    eq_to_hom (by rw [h₁, h₂]) ≫ (shift_add X i' j').hom :=
+by { cases h₁, cases h₂, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₁ (i i' j : A) (h : i = i') :
+  eq_to_hom (by rw h) ≫ (shift_add X i' j).inv = (shift_add X i j).inv ≫ eq_to_hom (by rw h) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₁]
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₂ (i j j' : A) (h : j = j') :
+  eq_to_hom (by rw h) ≫ (shift_add X i j').inv = (shift_add X i j).inv ≫ eq_to_hom (by rw h) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₂]
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₁₂ (i j i' j' : A) (h₁ : i = i') (h₂ : j = j') :
+  eq_to_hom (by rw [h₁, h₂]) ≫ (shift_add X i' j').inv =
+    (shift_add X i j).inv ≫ eq_to_hom (by rw [h₁, h₂]) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₁₂]
```

Comment:
I'm not a huge fan of the `_1` and `_2` subscripts in these lemmas. Can we come up with a more descriptive name?

Reply:
would `_l` `_r` or `_left` `_right` be better?